src/main/scala/mesosphere/marathon/api/LeaderProxyFilter.scala

```diff
@@ -1,137 +1,263 @@
+  }
+}
 
-              val proxyOutputStream = proxy.getOutputStream
-              copy(request.getInputStream, proxyOutputStream)
-              proxyOutputStream.close
-          }
+object LeaderProxyFilter {
+  private val log = Logger.getLogger(getClass.getName)
+  val ERROR_STATUS_NO_CURRENT_LEADER: String = "Could not determine the current leader"
+}
+
+/**
+  * Forwards a HttpServletRequest to an URL.
+  */
+trait RequestForwarder {
+  def forward(url: URL, request: HttpServletRequest, response: HttpServletResponse): Unit
+}
 
-          val status = proxy.getResponseCode
-          response.setStatus(status)
+class JavaUrlConnectionRequestForwarder @Inject() (
+  leaderProxyConf: LeaderProxyConf, @Named(ModuleNames.NAMED_HOST_PORT) myHostPort: String)
+    extends RequestForwarder {
 
-          val responseOutputStream = response.getOutputStream
+  import JavaUrlConnectionRequestForwarder._
 
-          try {
-            val fields = proxy.getHeaderFields
-            // getHeaderNames() and getHeaders() are known to return null
-            if (fields != null) {
-              for ((name, values) <- fields.asScala) {
-                if (name != null && values != null) {
-                  for (value <- values.asScala) {
-                    response.setHeader(name, value)
-                  }
-                }
-              }
-            }
-            copy(proxy.getInputStream, response.getOutputStream)
-            proxy.getInputStream.close()
-          }
-          catch {
-            case e: Exception =>
-              copy(proxy.getErrorStream, response.getOutputStream)
-              proxy.getErrorStream().close()
-          }
-          finally {
-            responseOutputStream.close()
-          }
+  private[this] val viaValue: String = s"1.1 $myHostPort"
+
+  override def forward(url: URL, request: HttpServletRequest, response: HttpServletResponse): Unit = {
+
+    val viaOpt = Option(request.getHeaders(HEADER_VIA)).map(_.asScala.toVector)
+    log.debug("viaOpt {}", viaOpt)
+    if (viaOpt.exists(_.contains(viaValue))) {
+      log.error("Prevent proxy cycle, rejecting request")
+      response.sendError(HttpStatus.SC_BAD_GATEWAY, ERROR_STATUS_LOOP)
+      return
+    }
+
+    val method = request.getMethod
+
+    log.info(s"Proxying request to $method $url from $myHostPort")
+
+    val leaderConnection: HttpURLConnection = url.openConnection().asInstanceOf[HttpURLConnection]
+
+    leaderConnection.setConnectTimeout(leaderProxyConf.leaderProxyConnectionTimeout())
+    leaderConnection.setReadTimeout(leaderProxyConf.leaderProxyReadTimeout())
+
+    // getHeaderNames() and getHeaders() are known to return null, see:
+    //http://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html#getHeaders(java.lang.String)
+    val names = Option(request.getHeaderNames).map(_.asScala).getOrElse(Nil)
+    for {
+      name <- names if !name.equalsIgnoreCase("host") && !name.equalsIgnoreCase("connection")
```

Comment:
Most of the time you use `proxy_set_header Host $host;` if upstream depends on host header somehow.